### PR TITLE
Ensure `watch_object` handles objects removed before init

### LINKED ISSUE #1576: Stream produced by "watch_object" no longer emits when an object isn't in the initial list
### Current and expected behavior

A test began failing after a Kube-RS upgrade from 0.91 to 0.93 - rough sketch:

````
let _ = api.delete(&obj).await;

// never returns now as there is no object at the point of awaiting
await_condition(self.api.clone(), obj.name_unchecked(), conditions::is_deleted).await;
````

Looking at the earlier implementation an `Ok(None)` event would be passed up to check when no objects were returned. - https://github.com/kube-rs/kube/blob/5dbae3a18c14a225d2d993b9effd16147fef420e/kube-runtime/src/watcher.rs#L815 

After 0.92 the "InitDone" always leads to `None` - https://github.com/kube-rs/kube/blob/0f6cb6f0ac695444f6789f98fa07073f4980a127/kube-runtime/src/watcher.rs#L830

### Possible solution

Use "scan" to track whether or not the object was seen for each round of `Init` events. If the object was seen `InitDone` will return `None` otherwise it should pass `Ok(None)`.

### Additional context

_No response_

### Environment

N/A

### Configuration and features

_No response_

### Affected crates

kube-runtime

### Would you like to work on fixing this bug?

yes

### PR DESCRIPTION
**Pull request by markdingram**
Stream produced by `watch_object` will include an item when the object isn't in any initial list

fixes #1576

## Solution

No tests! I see there are attempts already in this area underway, e.g. https://github.com/kube-rs/kube/pull/1529

I would be happy to add tests if/when those tests land

### CODE REVIEW

File: kube-runtime/src/watcher.rs [hunk 1]
```diff
@@ -844,18 +844,36 @@ pub fn watch_object<K: Resource + Clone + DeserializeOwned + Debug + Send + 'sta
...
+                // Pass up `None` for Deleted
+                Ok(Event::Delete(_)) => Some(Ok(None)),
+                // Ignore marker event
+                Ok(Event::Init) => None,
+                // Pass up `None` if the object wasn't seen in any initial list
+                Ok(Event::InitDone) => {
+                    if obj_seen {
+                        None
+                    } else {
+                        Some(Ok(None))
```
Comment (clux) [inline kube-runtime/src/watcher.rs:870]: So, if i understand this correctly, the idea here is that; if we didn't see the object in the `Init`, then that effectively means that the object was successfully deleted before we got to list, so there's no point waiting anymore?

File: kube-runtime/src/watcher.rs [hunk 2]
```diff
@@ -844,18 +844,36 @@ pub fn watch_object<K: Resource + Clone + DeserializeOwned + Debug + Send + 'sta
...
-            Err(err) => Some(Err(err)),
-        }
-    })
+    watcher(api, Config::default().fields(&fields))
+        // track whether the object was seen in each initial listing
+        .scan(false, |obj_seen, event| {
+            if matches!(event, Ok(Event::Init)) {
+                *obj_seen = false;
+            } else if matches!(event, Ok(Event::InitApply(_))) {
+                *obj_seen = true;
```
Comment (clux) [inline kube-runtime/src/watcher.rs:853]: so this branch basically denotes: "yes, we saw it, and we will be waiting for a delete to happen"

**Review by clux — COMMENTED** [review]
Thanks for this. I think this makes sense. A couple of quick questions to make sure i understand properly.

File: kube-runtime/src/watcher.rs (hunk 1 above)
Comment (markdingram) [reply kube-runtime/src/watcher.rs:870]: Yes! I'll add comments to make the logic clearer

**Review by clux — APPROVED** [review]
Comments are clear, thank you!

**Pull request merged**